src/lib.rs

```diff
@@ -14,46 +14,109 @@
+    Brotli,
+    /// No encoding is preferred.
+    Identity,
+    /// No preference is expressed on which encoding to use. Either the `Accept-Encoding` header is not present, or `*` is set as the most preferred encoding.
+    None,
+}
+
+impl Encoding {
+    /// Parses a given string into its corresponding encoding.
+    fn parse(s: &str) -> Result<Encoding> {
+        match s {
+            "gzip" => Ok(Encoding::Gzip),
+            "deflate" => Ok(Encoding::Deflate),
+            "br" => Ok(Encoding::Brotli),
+            "identity" => Ok(Encoding::Identity),
+            "*" => Ok(Encoding::None),
+            _ => Err(ErrorKind::UnknownEncoding)?,
+        }
+    }
+
+    /// Converts the encoding into its' corresponding header value.
+    ///
+    /// Note that [`Encoding::None`] will return a HeaderValue with the content `*`.
+    /// This is likely not what you want if you are using this to generate the `Content-Encoding` header to be included in an encoded response.
+    pub fn to_header_value(self) -> HeaderValue {
+        match self {
+            Encoding::Gzip => HeaderValue::from_str("gzip").unwrap(),
+            Encoding::Deflate => HeaderValue::from_str("deflate").unwrap(),
+            Encoding::Brotli => HeaderValue::from_str("br").unwrap(),
+            Encoding::Identity => HeaderValue::from_str("identity").unwrap(),
+            Encoding::None => HeaderValue::from_str("*").unwrap(),
+        }
+    }
 }
 
 /// Parse a set of HTTP headers into an `Encoding`.
 pub fn parse(headers: &HeaderMap) -> Result<Encoding> {
-  let header = match headers.get(ACCEPT_ENCODING) {
-    Some(header) => header,
-    None => return Ok(Encoding::None),
-  };
-
-  let string = header.to_str().context(ErrorKind::InvalidEncoding)?;
-
-  if string.contains("br") {
-    Ok(Encoding::Brotli)
-  } else if string.contains("deflate") {
-    Ok(Encoding::Deflate)
-  } else if string.contains("gzip") {
-    Ok(Encoding::Gzip)
-  } else {
-    Err(ErrorKind::UnknownEncoding)?
-  }
+    let mut preferred_encoding = Encoding::None;
+    let mut max_qval = 0.0;
+
+    for header_value in headers.get_all(ACCEPT_ENCODING).iter() {
+        let header_value = header_value.to_str().context(ErrorKind::InvalidEncoding)?;
+        for v in header_value.split(',').map(str::trim) {
+            let mut v = v.splitn(2, ";q=");
+            let encoding = v.next().unwrap();
+
+            match Encoding::parse(encoding) {
+                Ok(encoding) => {
+                    if let Some(qval) = v.next() {
+                        let qval = match qval.parse::<f32>() {
+                            Ok(f) => f,
+                            Err(_) => return Err(ErrorKind::InvalidEncoding)?,
+                        };
+                        if (qval - 1.0f32).abs() < 0.01 {
+                            preferred_encoding = encoding;
+                            break;
+                        } else if qval > 1.0 {
```

Comment:
Should we move this branch as the last branch?

I think error handling branches can come at last.

Reply:
It's a little bit difficult to track the comments here since this PR was merged already, but I'll reply here anyways.

Actually, there is a slip of logic here that would require the predicate there to actually come *first* instead of last (since qval can technically be `1.001` and it would still pass, even though ideally it should fail since a q-value shouldn't be larger than `1.0` flat).

This is actually addressed indirectly through the refactoring on the second PR I've made: https://github.com/rustasync/accept-encoding/pull/2

Since the logic in that function runs first anyway.